<?xml version="1.0"?>

<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="pan_tilt" description="iqr pan tilt">
  <xacro:property name="PI" value="3.1415926535897931" />
  <xacro:macro name="pan_tilt" params="parent *origin">
  <xacro:property name="prefix" value="pan_tilt" />
    <joint name="${prefix}_base_joint" type="fixed">
      <xacro:insert_block name="origin" />
      <parent link="${parent}"/>
      <child link="${prefix}_base_link"/>
    </joint>

    <link name="${prefix}_base_link">
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <geometry>
          <mesh filename="package://pan_tilt_description/urdf/mesh/base.stl" />
        </geometry>
        <material name="">
          <color rgba="0.2 0.2 0.2 1" />
        </material>
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <geometry>
          <mesh filename="package://pan_tilt_description/urdf/mesh/base.stl" />
        </geometry>
      </collision>  
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="1"/>
        <inertia ixx="0.01"  ixy="0"  ixz="0" iyy="0.01" iyz="0" izz="0.01" />
      </inertial>
    </link>
    <gazebo reference="${prefix}_base_link">
       <material>Gazebo/Gray</material>
    </gazebo>

    <link name="${prefix}_yaw_link">
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <geometry>
          <mesh filename="package://pan_tilt_description/urdf/mesh/pan.stl" />
        </geometry>
        <material name="">
          <color rgba="0.2 0.2 0.2 1" />
        </material>
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <geometry>
          <mesh filename="package://pan_tilt_description/urdf/mesh/pan.stl" />
        </geometry>
      </collision>  

      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="1"/>
        <inertia ixx="0.01"  ixy="0"  ixz="0" iyy="0.01" iyz="0" izz="0.01" />
      </inertial>
    </link>
    <gazebo reference="${prefix}_yaw_link">
       <material>Gazebo/Gray</material>
    </gazebo>

    <joint name="${prefix}_yaw_joint" type="revolute">
      <origin xyz="0.0 0.0 0.05455" rpy="0.0 0.0 0.0"/>
      <parent link="${prefix}_base_link"/>
      <child link="${prefix}_yaw_link"/>
      <axis xyz="0 0 1"/>
      <limit lower="-0.8" upper="0.4" effort="10.0" velocity="1.0"/>
    </joint>
    <transmission name="yaw_tran">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="${prefix}_yaw_joint">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </joint>
      <actuator name="${prefix}yaw_motor">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
    </transmission>

    <link name="${prefix}_pitch_link">
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <geometry>
          <mesh filename="package://pan_tilt_description/urdf/mesh/tilt.stl" />
        </geometry>
        <material name="">
          <color rgba="0.2 0.2 0.2 1" />
        </material>
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <geometry>
          <mesh filename="package://pan_tilt_description/urdf/mesh/tilt.stl" />
        </geometry>
      </collision>  

      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="1"/>
        <inertia ixx="0.01"  ixy="0"  ixz="0" iyy="0.01" iyz="0" izz="0.01" />
      </inertial>
    </link>
    <gazebo reference="${prefix}_pitch_link">
       <material>Gazebo/Gray</material>
    </gazebo>

    <joint name="${prefix}_pitch_joint" type="revolute">
      <origin xyz="0.0 0.0 0.057" rpy="0.0 0.0 0.0"/>
      <parent link="${prefix}_yaw_link"/>
      <child link="${prefix}_pitch_link"/>
      <axis xyz="0 1 0"/>
      <limit lower="-1.57" upper="1.57" effort="10.0" velocity="1.0"/>
    </joint>
    <transmission name="pitch_tran">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="${prefix}_pitch_joint">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </joint>
      <actuator name="${prefix}pitch_motor">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
    </transmission>

    <link name="${prefix}_kinect_surface">
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="1"/>
        <inertia ixx="0.01"  ixy="0"  ixz="0" iyy="0.01" iyz="0" izz="0.01" />
      </inertial>
      <collision>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <geometry >
          <box size="0.01 0.01 0.01" />
        </geometry>
      </collision>  
    </link>
    <gazebo reference="${prefix}_kinect_surface">
      <material>Gazebo/Gray</material>
    </gazebo>

    <joint name="${prefix}_kinect_surface_joint" type="fixed">
      <origin xyz="0.0010 0.0077 -0.022" rpy="0 0 0" />
      <parent link="${prefix}_pitch_link"/>
      <child link="${prefix}_kinect_surface"/>
    </joint>
  </xacro:macro>  
</robot>
